File: ConfigurationManager/Utils/Converter.cs
using System;
using System.ComponentModel;

namespace ConfigurationManager.Utils
{
    public static class Converter
    {
        public static T Convert<T>(this string value)
        {
            var converter = TypeDescriptor.GetConverter(typeof(T));
            return (T)converter.ConvertFromString(value);
        }

        public static object Convert(this string value, Type type)
        {
            var converter = TypeDescriptor.GetConverter(type);
            return converter.ConvertFromString(value);
        }

        public static bool TryConvert(this string value, Type type, out object convertedValue)
        {
            try
            {
                var converter = TypeDescriptor.GetConverter(type);
                convertedValue = converter.ConvertFromString(value);
                return true;
            }
            catch
            {
                convertedValue = null;
                return false;
            }
        }
    }
}

File: src/ConfigurationProvider/Converter.fs
module ConfigurationProvider.Converter

    open System
    open System.ComponentModel
    
    type String with
        member str.Convert (targetType: Type) =
            match str with 
                | null when targetType = typeof<string> ->
                    raise (NotSupportedException "Conversion of null to empty string is overriden")
                | x ->
                    let converter = TypeDescriptor.GetConverter targetType
                    converter.ConvertFromString x
        member str.Convert<'a>() =
            str.Convert typeof<'a> :?> 'a
            
        member str.TryConvert (targetType: Type) =
            match str with 
                | null when targetType.IsValueType ->
                    (false, Activator.CreateInstance(targetType))
                | x ->
                    try
                        let converter = TypeDescriptor.GetConverter targetType                    
                        (true, converter.ConvertFromString x)
                    with e ->
                        (false, null)
                        
        member str.TryConvert<'a> (targetType: Type) =
            let conversionResult = str.TryConvert targetType
            (fst conversionResult, snd conversionResult :?> 'a)
                